admiral/pkg/clusters/types.go

```diff
@@ -127,11 +130,185 @@ func (g *globalTrafficCache) Put(gtp *v1.GlobalTrafficPolicy) error {
+
+	mutex *sync.Mutex
+}
+
+
+func (r *routingPolicyCache) Delete(identity string, environment string) {
+	defer r.mutex.Unlock()
+	r.mutex.Lock()
+	key := common.ConstructRoutingPolicyKey(environment, identity)
+	if _, ok := r.identityCache[key]; ok {
+		log.Infof("deleting RoutingPolicy with key=%s from global RoutingPolicy cache", key)
+		delete(r.identityCache, key)
+	}
+}
+
+func (r *routingPolicyCache ) GetFromIdentity(identity string, environment string) *v1.RoutingPolicy {
+	defer r.mutex.Unlock()
+	r.mutex.Lock()
+	return r.identityCache[common.ConstructRoutingPolicyKey(environment, identity)]
+}
+
+func (r *routingPolicyCache) Put(rp *v1.RoutingPolicy) error {
+	if rp == nil || rp.Name == "" {
+		// no RoutingPolicy, throw error
+		return errors.New("cannot add an empty RoutingPolicy to the cache")
+	}
+	if rp.Labels == nil {
+		return errors.New("labels empty in RoutingPolicy")
+	}
+	defer r.mutex.Unlock()
+	r.mutex.Lock()
+	var rpIdentity = rp.Labels[common.GetRoutingPolicyLabel()]
+	var rpEnv = common.GetRoutingPolicyEnv(rp)
+
+	log.Infof("Adding RoutingPolicy with name %v to RoutingPolicy cache. LabelMatch=%v env=%v", rp.Name, rpIdentity, rpEnv)
+	key := common.ConstructRoutingPolicyKey(rpEnv, rpIdentity)
+	r.identityCache[key] = rp
+
+	return nil
+}
+
+
+type routingPolicyFilterCache struct {
+	// map of envoyFilters key=environment+identity of the routingPolicy, value is a map [clusterId -> map [filterName -> filterName]]
+	filterCache map[string]map[string]map[string]string
+	mutex *sync.Mutex
+}
+
+func (r *routingPolicyFilterCache) Get(identityEnvKey string) (filters map[string]map[string]string) {
+	defer r.mutex.Unlock()
+	r.mutex.Lock()
+	return r.filterCache[identityEnvKey]
+}
+
+func (r *routingPolicyFilterCache) Put(identityEnvKey string, clusterId string, filterName string) {
+	defer r.mutex.Unlock()
+	r.mutex.Lock()
+	if r.filterCache[identityEnvKey] == nil {
+		r.filterCache[identityEnvKey] = make(map[string]map[string]string)
+	}
+
+	if r.filterCache[identityEnvKey][clusterId] == nil {
+		r.filterCache[identityEnvKey][clusterId] = make(map[string]string)
+	}
+	r.filterCache[identityEnvKey][clusterId][filterName] = filterName
+}
+
+func (r *routingPolicyFilterCache) Delete(identityEnvKey string) {
+	defer r.mutex.Unlock()
+	r.mutex.Lock()
+	// delete all envoyFilters for a given identity+env key
+	delete(r.filterCache, identityEnvKey)
+}
+func (r RoutingPolicyHandler) Added(obj *v1.RoutingPolicy) {
+	if common.ShouldIgnoreResource(obj.ObjectMeta) {
+		log.Infof(LogErrFormat, "success", "routingpolicy", obj.Name, obj.ClusterName, "Ignored the RoutingPolicy because of the annotation")
+		return
+	}
+	dependents, done := getDependents(obj, r)
+	if done {
+		return
+	}
+	r.processroutingPolicy(dependents, obj, admiral.Add)
+	log.Info("finished processing routing policy")
+}
+
+func (r RoutingPolicyHandler) processroutingPolicy(dependents map[string]string, routingPolicy *v1.RoutingPolicy, eventType admiral.EventType ) {
```

Comment:
same thing here, can we return the `error` from this func rather than logging it here?
That way, line #268 can be wrapped in an `if err` block.

Reply:
done